### PR TITLE
Support Subjobs

### PR DESCRIPTION
**Pull request by Grillz**
This PR adds support for subjobs, which are needed when branch indexing is used.

### CODE REVIEW

File: jenkins.go [hunk 1]
```diff
@@ -302,6 +302,18 @@ func (j *Jenkins) GetJob(id string) (*Job, error) {
 	return nil, errors.New(strconv.Itoa(status))
 }
 
+func (j *Jenkins) GetSubJob(parentId string, childId string) (*Job, error) {
```
Comment (kris-nova) [inline jenkins.go]: Optional method signature change

```
func (j *Jenkins) GetSubJob(parentId, childId string) (*Job, error)
```

File: jenkins.go [hunk 2]
```diff
@@ -302,6 +302,18 @@ func (j *Jenkins) GetJob(id string) (*Job, error) {
...
+	job := Job{Jenkins: j, Raw: new(jobResponse), Base: "/job/" + parentId + "/job/" + childId}
```
Comment (kris-nova) [inline jenkins.go]: Do we need to validate the ID's at all? What if they are `""` or something? Maybe 

```
if len(parentId) < 1 {
    return nil, fmt.Errorf("Invalid ID")
}
```

File: jenkins.go [hunk 3]
```diff
@@ -302,6 +302,18 @@ func (j *Jenkins) GetJob(id string) (*Job, error) {
...
+	status, err := job.Poll()
+	if err != nil {
+		return nil, err
+	}
+	if status == 200 {
```
Comment (kris-nova) [inline jenkins.go]: Is this an HTTP status code? Code we have any 200 -> 299 level response and still be valid?

File: jenkins.go [hunk 4]
```diff
@@ -302,6 +302,18 @@ func (j *Jenkins) GetJob(id string) (*Job, error) {
...
+		return &job, nil
+	}
+	return nil, errors.New(strconv.Itoa(status))
```
Comment (kris-nova) [inline jenkins.go]: This string doesn't give the user much information on what went wrong, maybe 

```
return nil, fmt.Errorf("Invalid response code: %d", status)
```

File: jenkins.go [hunk 5]
```diff
@@ -302,6 +302,18 @@ func (j *Jenkins) GetJob(id string) (*Job, error) {
 	return nil, errors.New(strconv.Itoa(status))
 }
 
+func (j *Jenkins) GetSubJob(parentId string, childId string) (*Job, error) {
+	job := Job{Jenkins: j, Raw: new(jobResponse), Base: "/job/" + parentId + "/job/" + childId}
+	status, err := job.Poll()
+	if err != nil {
+		return nil, err
```
Comment (kris-nova) [inline jenkins.go]: Again here, maybe do the end user a favor and give them a little insight into the error?

```
return nil, fmt.Errorf("Unable to poll job: %v", err)
```

File: job.go [hunk 6]
```diff
@@ -179,6 +180,10 @@ func (j *Job) GetAllBuildIds() ([]jobBuild, error) {
 	return buildsResp.Builds, nil
 }
 
+func (j *Job) GetSubJobsMetadata() []job {
+	return j.Raw.SubJobs
```
Comment (kris-nova) [inline job.go]: Error if the slice is empty or even initialized? 

```
func (j *Job) GetSubJobsMetadata() ([]job, error) {
    if len(j.Raw.SubJobs) > 0 {
        return j.Raw.SubJobs, nil
    }
    var er []job
    return er, fmt.Errorf("No jobs fond")
}
```

File: job.go [hunk 7]
```diff
@@ -187,6 +192,18 @@ func (j *Job) GetDownstreamJobsMetadata() []job {
 	return j.Raw.DownstreamProjects
 }
 
+func (j *Job) GetSubJobs() ([]*Job, error) {
+	jobs := make([]*Job, len(j.Raw.SubJobs))
+	for i, job := range j.Raw.SubJobs {
+		ji, err := j.Jenkins.GetSubJob(j.GetName(), job.Name)
+		if err != nil {
+			return nil, err
```
Comment (kris-nova) [inline job.go]: Context to error here? Examples above.

File: jenkins.go (hunk 1 above)
Comment (Grillz) [inline jenkins.go]: Personally not a fan of the implied types

File: jenkins.go (hunk 2 above)
Comment (Grillz) [inline jenkins.go]: I am going to assume the user isn't that senseless

File: jenkins.go (hunk 5 above)
Comment (Grillz) [inline jenkins.go]: That seems plenty insight to me ;)

File: jenkins.go (hunk 3 above)
Comment (Grillz) [inline jenkins.go]: Jenkins is currently only sending back 200s out of their API, but this could protect future breaks

File: jenkins.go (hunk 4 above)
Comment (Grillz) [inline jenkins.go]: again a weakness in the Jenkins API, really not the best. There would have to be a chart of error codes to their mappings

File: job.go (hunk 6 above)
Comment (Grillz) [inline job.go]: Empty array is the flow of the rest of the project

File: job.go (hunk 7 above)
Comment (Grillz) [inline job.go]: Context is already happening in the method

File: job.go [hunk 8]
```diff
@@ -78,6 +78,7 @@ type jobResponse struct {
 	LastUnstableBuild     jobBuild `json:"lastUnstableBuild"`
 	LastUnsuccessfulBuild jobBuild `json:"lastUnsuccessfulBuild"`
 	Name                  string   `json:"name"`
+	SubJobs               []job    `json:"jobs"`
```
Comment (supereagle) [inline job.go]: Maybe it is better to keep the field names of Go object and its JSON object are the same.

**Comment by bndr** [comment]
Thanks for the review guys! looks good to me, would appreciate some tests for this PR though.

**Comment by Grillz** [comment]
@bndr caught me ;) I'll put those in tonight

**Comment by kris-nova** [comment]
So happy to see this merged, excellent work by @Grillz <3

**Comment by bndr** [comment]
Sorry it took so long :)

**Pull request merged**